### PR TITLE
Implemented partial date/time in ITEM_STRUCT

### PR DESCRIPTION
**Pull request by chevalleyc**
Changes db encoding with support for partial date time. 

See https://github.com/ehrbase/openEHR_SDK/blob/feature/197_partial_ISO_strings/serialisation/src/test/java/org/ehrbase/serialisation/dbencoding/DBEncodeTest.java testDateTimeEncodeDecode() with composition example: https://github.com/ehrbase/openEHR_SDK/blob/feature/197_partial_ISO_strings/test-data/src/main/resources/composition/canonical_json/datetime_tests.json

This implementation requires https://github.com/openEHR/archie/pull/208 to perform as expected

LIMITATIONS:
1. This changes are for the standard ISO8601 date/time representation. The so-called 'compact' is not serviced by this PR.
2. Partial time limited to hours always come back with minutes. F.e `10` is defaulted to `10:00`. This is due to a Java API limitation.
3. This changes are for DV_TEMPORAL in ITEM_STRUCTURE (that is, all temporal held in the JSON encoded structure). We are evaluating the need to propagate these changes to date/time in EVENT_CONTEXT and PARTICIPATION assuming this makes sense to handle partial date/time there.

### CODE REVIEW

**Review by Geisterfalle — COMMENTED** [review]
I checked this one with an diagnosis example. Based on your example composition I would assume that it should work, however, I get the following error related to archie:

`{
    "error": "Text '2020-09-02' could not be parsed at index 4:2020-09-02 (through reference chain: com.nedap.archie.rm.composition.Composition[\"content\"]->java.util.ArrayList[0]->com.nedap.archie.rm.composition.Evaluation[\"data\"]->com.nedap.archie.rm.datastructures.ItemTree[\"items\"]->java.util.ArrayList[1]->com.nedap.archie.rm.datastructures.Element[\"value\"]->com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime[\"value\"])",
    "status": "Bad Request"
}`

I understand that required changes in Archie are already merged. Am I missing something, maybe in how I need to set up EHRbase to properly test?

**Comment by chevalleyc** [comment]
> I checked this one with an diagnosis example. Based on your example composition I would assume that it should work, however, I get the following error related to archie:
> 
> `{ "error": "Text '2020-09-02' could not be parsed at index 4:2020-09-02 (through reference chain: com.nedap.archie.rm.composition.Composition[\"content\"]->java.util.ArrayList[0]->com.nedap.archie.rm.composition.Evaluation[\"data\"]->com.nedap.archie.rm.datastructures.ItemTree[\"items\"]->java.util.ArrayList[1]->com.nedap.archie.rm.datastructures.Element[\"value\"]->com.nedap.archie.rm.datavalues.quantity.datetime.DvDateTime[\"value\"])", "status": "Bad Request" }`
> 
> I understand that required changes in Archie are already merged. Am I missing something, maybe in how I need to set up EHRbase to properly test?

The changes are merged I believe, however make sure you rebuild everything from scratch (mvn clean), and check your archie version in your .m2 directory. I found mine was not properly updated for some reason hence causing the same issue.

File: serialisation/src/test/java/org/ehrbase/serialisation/dbencoding/DBEncodeTest.java [hunk 1]
```diff
@@ -283,6 +283,38 @@ public void testDvIntervalRoundTrip() throws IOException {
         assertNotNull(composition2);
     }
 
+    @Test
+    public void testDateTimeEncodeDecode() throws IOException {
+        Composition composition = new CanonicalJson().unmarshal(IOUtils.resourceToString("/composition/canonical_json/datetime_tests.json", UTF_8), Composition.class);
```
Comment (stefanspiska) [inline serialisation/src/test/java/org/ehrbase/serialisation/dbencoding/DBEncodeTest.java:288]: Please add datetime_tests.json to CompositionTestDataCanonicalJson and use that

File: serialisation/src/test/java/org/ehrbase/serialisation/jsonencoding/CanonicalJsonMarshallingTest.java [hunk 2]
```diff
@@ -45,4 +47,26 @@ public void UnmarshalItemTree() throws IOException {
...
+
+    @Test
+    public void UnmarshalPartialDate() throws IOException {
+
+        String value = new String(Files.readAllBytes(Paths.get("src/test/resources/sample_data/partialdvdate.json")));
+
+        CanonicalJson cut = new CanonicalJson();
+        DvDate dvDate = cut.unmarshal(value, DvDate.class);
+
+        assertNotNull(dvDate);
```
Comment (stefanspiska) [inline serialisation/src/test/java/org/ehrbase/serialisation/jsonencoding/CanonicalJsonMarshallingTest.java:59]: maybe test the value

File: serialisation/src/test/java/org/ehrbase/serialisation/jsonencoding/CanonicalJsonMarshallingTest.java [hunk 3]
```diff
@@ -45,4 +47,26 @@ public void UnmarshalItemTree() throws IOException {
...
+
+    @Test
+    public void UnmarshalPartialDateTime() throws IOException {
+
+        String value = new String(Files.readAllBytes(Paths.get("src/test/resources/sample_data/partialdvdatetime.json")));
+
+        CanonicalJson cut = new CanonicalJson();
+        DvDateTime dvDateTime = cut.unmarshal(value, DvDateTime.class);
+
+        assertNotNull(dvDateTime);
```
Comment (stefanspiska) [inline serialisation/src/test/java/org/ehrbase/serialisation/jsonencoding/CanonicalJsonMarshallingTest.java:70]: maybe test the value

**Review by stefanspiska — CHANGES_REQUESTED** [review]
Could you pleas use the enums in test-data module for test compositions.

Otherwise it looks good.

**Pull request merged**